File: src/main/java/me/saro/kit/webs/WebResult.java
package me.saro.kit.webs;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * web result
 * @author PARK Yong Seo
 * @since 1.0.0
 */
public class WebResult {

    WebResult() {
    }

    // http status
    private int status = -1;

    // exception
    private Exception exception;

    // headers
    private Map<String, List<String>> headers = Collections.emptyMap();

    // response body data
    private String body;

    /**
     * is status 2xx + have not exception
     * @return
     */
    public boolean isSuccess() {
        return isStatus2xx() && exception == null;
    }
    
    /**
     * is status 2xx
     * @return
     */
    public boolean isStatus2xx() {
        return status >= 200 && status < 300;
    }

    /**
     * is status 3xx
     * @return
     */
    public boolean isStatus3xx() {
        return status >= 300 && status < 400;
    }

    /**
     * is status 4xx
     * @return
     */
    public boolean isStatus4xx() {
        return status >= 400 && status < 500;
    }

    /**
     * is status 5xx
     * @return
     */
    public boolean isStatus5xx() {
        return status >= 500 && status < 600;
    }
    
    /**
     * has body
     * @return
     */
    public boolean hasBody() {
        return body != null;
    }

    public Optional<String> getBody() {
        return Optional.ofNullable(body);
    }

    public String getBody(String orElse) {
        return body != null ? body : orElse;
    }

    public int getStatus() {
        return status;
    }

    public void setStatus(int status) {
        this.status = status;
    }

    public Exception getException() {
        return exception;
    }

    public void setException(Exception exception) {
        this.exception = exception;
    }

    public Map<String, List<String>> getHeaders() {
        return headers;
    }

    public void setHeaders(Map<String, List<String>> headers) {
        this.headers = headers;
    }

    public void setBody(String body) {
        this.body = body;
    }

    @Override
    public String toString() {
        return "WebResult{" +
                "status=" + status +
                ", exception=" + exception +
                ", headers=" + headers +
                ", body=" + body +
                '}';
    }
}

File: src/test/java/me/saro/test/legacy/webs/WebResult.kt
package me.saro.test.legacy.webs

import java.util.*
import java.util.function.Predicate

/**
 * web result
 * @author PARK Yong Seo
 * @since 1.0.0
 */
class WebResult internal constructor() {
    // http status
    var status = -1

    // exception
    var exception: Exception? = null

    // headers
    var headers = emptyMap<String, List<String>>()

    // response body data
    private var body: String? = null

    /**
     * is status 2xx + have not exception
     * @return
     */
    val isSuccess: Boolean
        get() = isStatus2xx && exception == null

    /**
     * is status 2xx
     * @return
     */
    val isStatus2xx: Boolean
        get() = status >= 200 && status < 300

    /**
     * is status 3xx
     * @return
     */
    val isStatus3xx: Boolean
        get() = status >= 300 && status < 400

    /**
     * is status 4xx
     * @return
     */
    val isStatus4xx: Boolean
        get() = status >= 400 && status < 500

    /**
     * is status 5xx
     * @return
     */
    val isStatus5xx: Boolean
        get() = status >= 500 && status < 600

    /**
     * has body
     * @return
     */
    fun hasBody(): Boolean {
        return body != null
    }

    fun getBody(): Optional<String> {
        return Optional.ofNullable(body)
    }

    fun getBody(filter: Predicate<WebResult?>): Optional<String> {
        return Optional.ofNullable(if (filter.test(this)) body else null)
    }

    fun getBody(orElse: String?): String {
        return if (body != null) body!! else orElse!!
    }

    fun setBody(body: String?) {
        this.body = body
    }

    override fun toString(): String {
        return "WebResult{" +
                "status=" + status +
                ", exception=" + exception +
                ", headers=" + headers +
                ", body=" + body +
                '}'
    }
}